flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5298,6 +5554,104 @@ class ArrayExprLowering {
     };
   }
 
+  /// Given an optional fir.box, returns an fir.box that is the original one if
+  /// it is present and it otherwise an unallocated box.
+  /// Absent fir.box are implemented as a null pointer descriptor. Generated
+  /// code may need to unconditionally read a fir.box that can be absent.
+  /// This helper allows creating a fir.box that can be read in all cases
+  /// outside of a fir.if (isPresent) region. However, the usages of the value
+  /// read from such box should still only be done in a fir.if(isPresent).
+  static fir::ExtendedValue
+  absentBoxToUnalllocatedBox(fir::FirOpBuilder &builder, mlir::Location loc,
+                             const fir::ExtendedValue &exv,
+                             mlir::Value isPresent) {
+    mlir::Value box = fir::getBase(exv);
+    mlir::Type boxType = box.getType();
+    assert(boxType.isa<fir::BoxType>() && "argument must be a fir.box");
+    mlir::Value emptyBox =
+        fir::factory::createUnallocatedBox(builder, loc, boxType, llvm::None);
+    mlir::Value safeToReadBox =
+        builder.create<mlir::SelectOp>(loc, isPresent, box, emptyBox);
+    return fir::substBase(exv, safeToReadBox);
+  }
+
+  CC genOptionalArr(const Fortran::lower::SomeExpr &expr) {
```

Comment:
Need some documentation here. I believe this is only ever called when the array expression is a whole reference appearing in an argument position in a procedure call. Is that right? If so, maybe call this something like `genarrForwardOptionalArgumentToCall`?

`genarr` is shorthand for "generate the code to possibly copy-in/copy-out an array and return the element by element code to be inserted in the subsequent loop nest". I could've camelCased it and typed it out everywhere, but that seemed ridiculous.